database/director.go

```diff
@@ -0,0 +1,150 @@
+package database
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/jellydator/ttlcache/v3"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/bcrypt"
+	"gorm.io/gorm"
+)
+
+var DirectorDB *gorm.DB
+
+type (
+	ApiKey struct {
+		ID          string `gorm:"primaryKey;column:id;type:text;not null;unique"`
+		Name        string `gorm:"column:name;type:text"`
+		HashedValue string `gorm:"column:hashed_value;type:text;not null"`
+		Scopes      string `gorm:"column:scopes;type:text"`
+		ExpiresAt   time.Time
+		CreatedAt   time.Time
+		CreatedBy   string `gorm:"column:created_by;type:text"`
+	}
+
+	ApiKeyCached struct {
+		Token        string // "$ID.$SECRET_IN_HEX" string form
+		Capabilities []string
+	}
+)
+
+func generateSecret(length int) ([]byte, error) {
+	bytesSlice := make([]byte, length)
+	_, err := rand.Read(bytesSlice)
+	if err != nil {
+		return nil, err
+	}
+	return bytesSlice, nil
+}
+
+func generateTokenID(secret []byte) string {
+	hash := sha256.Sum256(secret)
+	return hex.EncodeToString(hash[:])[:5]
+}
+
+func VerifyApiKey(apiKey string, verifiedKeysCache *ttlcache.Cache[string, ApiKeyCached]) (bool, []string, error) {
+	parts := strings.Split(apiKey, ".")
+	if len(parts) != 2 {
+		return false, nil, errors.New("invalid API key format")
+	}
+	id := parts[0]
+	secretHex := parts[1]
+
+	item := verifiedKeysCache.Get(id)
+	if item != nil {
+		// Cache hit
+		return true, item.Value().Capabilities, nil
```

Comment:
Missing the check that the secret is correct, no?  Seems to only verify the public portion here